cou/steps/analyze.py

```diff
@@ -0,0 +1,200 @@
+        logging.info(
+            "Charms are in the same openstack version and can be upgrade from: %s to: %s",
+            actual_release,
+            next_release,
+        )
+        upgrade_charms[next_release].update(os_versions.keys())
+
+    return upgrade_units, upgrade_charms
+
+
+def check_os_channels_and_migration(os_versions: Dict[str, defaultdict[Any, Set]]) -> Tuple:
+    """Check if applications has the right channel set and if needs charmhub migration."""
+    change_channel = defaultdict(set)
+    charmhub_migration = defaultdict(set)
+
+    for app in os_versions.keys():
+        app_channel, charm_origin = extract_app_channel_and_origin(app)
+        actual_release = list(os_versions[app].keys())
+        if len(actual_release) > 1:
+            logging.warning(
+                "Skip check of channels. App %s has units with different openstack version",
+                app,
+            )
+        else:
+            actual_release = actual_release[0]
+            expected_channel = f"{actual_release}/stable"
+            if actual_release not in app_channel:
+                change_channel[expected_channel].add(app)
+                logging.warning("App:%s need to track the channel: %s", app, expected_channel)
+            if charm_origin == "cs":
+                charmhub_migration[expected_channel].add(app)
+                logging.warning("App:%s need to perform migration to charmhub", app)
+
+    return change_channel, charmhub_migration
+
+
+def check_os_origin(os_versions: Dict[str, defaultdict[Any, Set]]) -> defaultdict[Any, Any]:
+    """Check if application config for openstack-origin is set right."""
+    change_openstack_release = defaultdict(set)
+    for app in os_versions.keys():
+        os_charm_config = extract_os_charm_config(app)
+        actual_release = list(os_versions[app].keys())
+
+        if len(actual_release) > 1:
+            logging.warning(
+                "Skip openstack-origin check. App %s has units with different openstack version",
+                app,
+            )
+
+        else:
+            actual_release = actual_release[0]
+            expected_os_origin = f"cloud:focal-{actual_release}"
+            # Exceptionally, if upgrading from Ussuri to Victoria
+            if actual_release == "ussuri":
+                if os_charm_config != "distro":
+                    logging.warning(
+                        "App: %s need to set openstack-origin or source to 'distro'", app
+                    )
+                    change_openstack_release["distro"].add(app)
+            else:
+                if expected_os_origin not in os_charm_config:
+                    change_openstack_release[expected_os_origin].add(app)
+                    logging.warning(
+                        "App: %s need to set openstack-origin or source to %s",
+                        app,
+                        expected_os_origin,
+                    )
```

Comment:
What is purpose to have one if condition in if condition?
```suggestion
            if actual_release == "ussuri" and os_charm_config != "distro":
                logging.warning("App: %s need to set openstack-origin or source to 'distro'", app)
                change_openstack_release["distro"].add(app)
            elif expected_os_origin not in os_charm_config:
                change_openstack_release[expected_os_origin].add(app)
                 logging.warning(
                    "App: %s need to set openstack-origin or source to %s",
                    app,
                    expected_os_origin,
                )
```